### ISSUE TITLE
exec on Windows doesn't find `.bat` files in PATH

### ISSUE THREAD

**Issue by deivid-rodriguez** [jruby/jruby#6745]
I _guess_ this might affect the whole `Process.spawn` family? Not really sure, I only tried `Kernel.exec`.

Rubygems installs `.bat` wrappers for gem executables on Windows to compensate for the lack of support for shebangs. They look like this:

```
@ECHO OFF
@"%~dp0ruby.exe" "%~dpn0" %*
```

On CRuby + Windows, `Kernel.exec` properly detects and uses these scripts:

```

PS V:\> ruby -e "Kernel.exec 'rake'"
-e:1:in `exec': No such file or directory - rake (Errno::ENOENT)
        from -e:1:in `<main>'

PS V:\> bundle
Fetching gem metadata from https://rubygems.org/
Fetching gem metadata from https://rubygems.org/.
Fetching rake 13.0.3
Installing rake 13.0.3
Using bundler 2.3.0.dev
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

PS V:\> ruby -e "Kernel.exec 'rake'"
rake aborted!
No Rakefile found (looking for: rakefile, Rakefile, rakefile.rb, Rakefile.rb)

(See full trace by running task with --trace)

PS V:\> rm C:\Ruby30-x64\bin\rake.bat

PS V:\> ruby -e "Kernel.exec 'rake'"
-e:1:in `exec': No such file or directory - rake (Errno::ENOENT)
        from -e:1:in `<main>'
```

However, on JRruby + Windows, they seem to be ignored:

```
PS V:\> jruby -e "Kernel.exec 'rake'"
Errno::ENOENT: No such file or directory - rake
  _exec_internal at org/jruby/RubyKernel.java:1767
            exec at uri:classloader:/jruby/kernel/kernel.rb:4
          <main> at -e:1

PS V:\> bundle
Fetching gem metadata from https://rubygems.org/
Fetching gem metadata from https://rubygems.org/.
Fetching rake 13.0.3
Installing rake 13.0.3
Using bundler 2.3.0.dev
Bundle complete! 1 Gemfile dependency, 2 gems now installed.
Use `bundle info [gemname]` to see where a bundled gem is installed.

PS V:\> jruby -e "Kernel.exec 'rake'"
Errno::ENOENT: No such file or directory - No message available
  _exec_internal at org/jruby/RubyKernel.java:1767
            exec at uri:classloader:/jruby/kernel/kernel.rb:4
          <main> at -e:1
```

**Comment by headius** [comment]
I feel like this one came up previously, but it is a good find nonetheless! This likely indicates that 9.2 is not searching for .bat files when looking for executables to run. I will look into it and try to find the other issue about this, which may be fixed on master (9.3).

**Comment by headius** [comment]
Previous report is #6516, supposedly fixed by #6532, but that involved `popen` which uses somewhat different logic. A similar fix may be appropriate here for `exec`.

**Comment by headius** [comment]
I think I see a possible cause. The logic for `exec` will attempt to do a real native exec, replacing the current process. On Windows, this logic is a bit different, using `spawn` or `aspawn` to overlay a new process over the current one. If that native logic fails, we fall back on older logic that never got the treatment from #6532.

https://github.com/jruby/jruby/blob/479c451181b7697eb9229017d3e2c4ef940ee27c/core/src/main/java/org/jruby/RubyKernel.java#L1848-L1867

This should be fixable by making the same changes as #6532 along the non-native exec path, but I will also investigate why this native exec is failing in some Windows environments.

**Comment by headius** [comment]
I suspect the reason this works without a vendored bundle dir is because JRuby will see the pure-Ruby `rake` script shipped in JRuby's bin dir and execute it directly. When there's only a .bat file we do not know what to execute because the `exec` logic does not consider .bat files as executable.

I have a vague recollection about modifying or configuring RubyGems on Windows to always install *both* the .bat wrapper and the pure-Ruby script, but I don't know that anything came of that.

**Comment by headius** [comment]
> always install both the .bat wrapper and the pure-Ruby script

FWIW this would also improve the performance of `bundle exec` on Windows since it would be able to see that the target command is a Ruby script. I don't think it can detect that the .bat wrapper would just run Ruby code, so it has to follow this exec path and start a new process.

**Comment by headius** [comment]
Confirmed that native exec is attempted but fails:

```
C:\jruby>bundle exec rake
native exec
failed
bundler: command not found: rake
Install missing gem executables with `bundle install`
```

**Comment by headius** [comment]
I have identified two issues I believe are causing this.

1. The native exec logic does work correctly, but due to this command line ("rake") having a single ellement we always run it through cmd.exe. Unfortunately the logic for native exec does not try to figure out the full path to cmd.exe, and unlike the other process-launching logic (which uses ProcessBuilder) this is needed for the call to CreateProcess.
2. The rake executable is also not resolving to rake.bat due to a bug in the executable verification code. The path is searched, and .bat extension tried, but the resulting path does not get used with cmd.exe and it fails anyway.

So this issue was triggered by the addition of native exec support, but only because it has stricter executable path requirements that did not exist in our exec preamble logic. Other process-launching paths should be ok for this case.

**Comment by headius** [comment]
I believe there is an issue with the generated .bat file as well.

This first snippit is the contents of the wrapper we ship in the bin/ dir. Note that it runs JRuby directly, using -S to launch the appropriate rake executable and `%*` to pass all remaining arguments to it.

```
@ECHO OFF
@"%~dp0jruby.exe" -S rake %*
```

The content below is what ends up in the gems/shared/bin/rake.bat file generated by our build when setting up the preinstalled gems:

```
@ECHO OFF
IF NOT "%~f0" == "~f0" GOTO :WinNT
@"java -cp ;C:\Users\Charles Oliver Nutte\.m2\repository\io\takari\polyglot\polyglot-ruby\0.3.1.1\polyglot-ruby-0.3.1.1.jar;C:\Users\Charles Oliver Nutte\.m2\repository\io\takari\polyglot\polyglot-common\0.3.1\polyglot-common-0.3.1.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\jruby\9.1.17.0\jruby-9.1.17.0.pom;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\jruby-core\9.1.17.0\jruby-core-9.1.17.0.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-netdb\1.1.6\jnr-netdb-1.1.6.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-enxio\0.16\jnr-enxio-0.16.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-x86asm\1.0.2\jnr-x86asm-1.0.2.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-unixsocket\0.17\jnr-unixsocket-0.17.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-posix\3.0.44\jnr-posix-3.0.44.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-constants\0.9.9\jnr-constants-0.9.9.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jffi\1.2.16\jffi-1.2.16.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jffi\1.2.16\jffi-1.2.16-native.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\joni\joni\2.1.15\joni-2.1.15.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\extras\bytelist\1.0.15\bytelist-1.0.15.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\jcodings\jcodings\1.0.27\jcodings-1.0.27.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\dirgra\0.3\dirgra-0.3.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\headius\invokebinder\1.10\invokebinder-1.10.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\headius\options\1.4\options-1.4.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\headius\unsafe-fences\1.0\unsafe-fences-1.0.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\jcraft\jzlib\1.1.3\jzlib-1.1.3.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\martiansoftware\nailgun-server\0.9.1\nailgun-server-0.9.1.jar;C:\Users\Charles Oliver Nutte\.m2\repository\joda-time\joda-time\2.8.2\joda-time-2.8.2.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\headius\modulator\1.0\modulator-1.0.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\jruby-stdlib\9.1.17.0\jruby-stdlib-9.1.17.0.jar org.jruby.Main" "C:/jruby/lib/ruby/gems/shared/bin/rake" %1 %2 %3 %4 %5 %6 %7 %8 %9
GOTO :EOF
:WinNT
@"java -cp ;C:\Users\Charles Oliver Nutte\.m2\repository\io\takari\polyglot\polyglot-ruby\0.3.1.1\polyglot-ruby-0.3.1.1.jar;C:\Users\Charles Oliver Nutte\.m2\repository\io\takari\polyglot\polyglot-common\0.3.1\polyglot-common-0.3.1.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\jruby\9.1.17.0\jruby-9.1.17.0.pom;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\jruby-core\9.1.17.0\jruby-core-9.1.17.0.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-netdb\1.1.6\jnr-netdb-1.1.6.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-enxio\0.16\jnr-enxio-0.16.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-x86asm\1.0.2\jnr-x86asm-1.0.2.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-unixsocket\0.17\jnr-unixsocket-0.17.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-posix\3.0.44\jnr-posix-3.0.44.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jnr-constants\0.9.9\jnr-constants-0.9.9.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jffi\1.2.16\jffi-1.2.16.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\github\jnr\jffi\1.2.16\jffi-1.2.16-native.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\joni\joni\2.1.15\joni-2.1.15.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\extras\bytelist\1.0.15\bytelist-1.0.15.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\jcodings\jcodings\1.0.27\jcodings-1.0.27.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\dirgra\0.3\dirgra-0.3.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\headius\invokebinder\1.10\invokebinder-1.10.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\headius\options\1.4\options-1.4.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\headius\unsafe-fences\1.0\unsafe-fences-1.0.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\jcraft\jzlib\1.1.3\jzlib-1.1.3.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\martiansoftware\nailgun-server\0.9.1\nailgun-server-0.9.1.jar;C:\Users\Charles Oliver Nutte\.m2\repository\joda-time\joda-time\2.8.2\joda-time-2.8.2.jar;C:\Users\Charles Oliver Nutte\.m2\repository\com\headius\modulator\1.0\modulator-1.0.jar;C:\Users\Charles Oliver Nutte\.m2\repository\org\jruby\jruby-stdlib\9.1.17.0\jruby-stdlib-9.1.17.0.jar org.jruby.Main" "%~dpn0" %*
```

Note that this is not even using current JRuby and is instead generating a `java` command line that uses **JRuby 9.1.17.0**. I believe this is because the gem install is run with the **jar** form of JRuby used in our build process. That form expects any future invocation will also be via a `java` command line with the JRuby jar files, so it generates this improper .bat file.

To make matters worse, this command line is improperly formatted. The `java` command **and** the classpath arguments **and** the `org.jruby.Main` are all within quotes, causing Windows to try to run that entire thing as an executable, failing with "The system cannot find the path specified." Splitting up that block into appropriate pieces allows it to run, but then it can't find the installed `rake` because no JRUBY_HOME env or jruby.home property are provided. This .bat file is just completely broken.

As a result, even with the two issues mentioned above fixed, it still fails due to the bat .bat script having a bad `java` command line.

**Comment by headius** [comment]
Another discovery... the RubyGems shipped with 9.1.17.0 generates these bat files differently from more recent versions.

```ruby
  def windows_stub_script(bindir, bin_file_name)
    ruby = Gem.ruby.gsub(/^\"|\"$/, "").tr(File::SEPARATOR, "\\")
    return <<-TEXT
@ECHO OFF
IF NOT "%~f0" == "~f0" GOTO :WinNT
@"#{ruby}" "#{File.join(bindir, bin_file_name)}" %1 %2 %3 %4 %5 %6 %7 %8 %9
GOTO :EOF
:WinNT
@"#{ruby}" "%~dpn0" %*
TEXT
  end
```

Instead of using the rbconfig `RUBY_INSTALL_NAME` the `Gem.ruby` function uses `ruby_install_name`. I may be able to trick this by overriding both of them but the root problem here is installing the gems from within the JRuby 9.1.17.0 jar file.

This additional issue with the .bar file may only affect **local builds** on Windows, since other platforms will not generate this file on install. I do not see it in the Linux-built JRuby 9.2.19.0 tarball, for example.

**Comment by headius** [comment]
The change is merged to 9.2 and will be merged to master shortly. A test was added that will run on master.

We will rip out this nasty code for 9.4 and replace it with something manageable.

**Comment by deivid-rodriguez** [comment]
Thanks for the fix @headius!
